library/std/src/net/ip/network.rs

```diff
@@ -0,0 +1,391 @@
+use super::{Ipv4Addr, Ipv6Addr};
+use crate::error::Error;
+use crate::fmt;
+
+/// Represents an [IPv4 address] prefix, describing a network of addresses.
+///
+/// An IPv4 address prefix can be represented by an IPv4 address and a prefix length.
+/// The prefix length specifies how many of the leftmost contiguous bits of the
+/// address comprise the prefix. The prefix then describes the network containing all
+/// addresses that start with the same bits as the prefix.
+///
+/// The size of an `Ipv6AddrPrefix` struct may vary depending on the target operating system.
+///
+/// # Textual representation
+///
+/// `Ipv4AddrPrefix` provides a [`FromStr`](crate::str::FromStr) implementation.
+/// The textual representation of an address prefix is the representation of an
+/// IPv4 address, followed by a separating `/` and then the prefix length in decimal notation;
+/// for example `192.0.2.0/24`.
+///
+///
+/// # Examples
+/// ```
+/// #![feature(ip_prefix)]
+///
+/// use std::net::{Ipv4Addr, Ipv4AddrPrefix};
+///
+/// // Create the address prefix `192.0.2.0/24`.
+/// let prefix = Ipv4AddrPrefix::new(Ipv4Addr::new(192, 0, 2, 0), 24).unwrap();
+///
+/// // The prefix is uniquely defined by a prefix address and prefix length.
+/// assert_eq!(prefix.address(), Ipv4Addr::new(192, 0, 2, 0));
+/// assert_eq!(prefix.len(), 24);
+///
+/// // The prefix describes the network of all addresses that start
+/// // with the same bits as the prefix.
+/// assert_eq!(prefix.contains(&Ipv4Addr::new(192, 0, 2, 0)), true);
+/// assert_eq!(prefix.contains(&Ipv4Addr::new(192, 0, 2, 7)), true);
+/// assert_eq!(prefix.contains(&Ipv4Addr::new(192, 0, 2, 255)), true);
+/// assert_eq!(prefix.contains(&Ipv4Addr::new(192, 0, 3, 0)), false);
+/// ```
+///
+/// [IPv4 address]: Ipv4Addr
+#[derive(Copy, PartialEq, Eq, Clone, Hash)]
+#[unstable(feature = "ip_prefix", issue = "86991")]
+pub struct Ipv4AddrPrefix {
+    address_raw: u32,
```

Comment:
Changed from `Ipv4Addr` to a `u32` to avoid storing potential extra fields of `c::in_addr`. Same thing for `Ipv6AddrPrefix`. This could be reverted if #78802 ever gets merged.